### PR TITLE
Make check_large_grads non-fatal

### PR DESCRIPTION
**Pull request by kwyss-nvidia**
Large gradient spikes may indicate a fault in a node, but they may be accurate based on model and data. Capture the signal with logging, but don't treat them as fatal with checkpoint and exit.

# What does this PR do ?
Allows training to continue when a gradient spike is checked.

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

**Comment by kwyss-nvidia** [comment]
Tagging @deepakn94

File: megatron/core/rerun_state_machine.py [hunk 1]
```diff
@@ -330,6 +330,17 @@ def train_step(data_iterator, ...):
                 self._maybe_report_stats()
                 self.saved_results = defaultdict(list)
                 return False
+            will_continue_tensor: torch.Tensor = torch.tensor(
```
Comment (deepakn94) [inline megatron/core/rerun_state_machine.py]: Nit: can you create a helper method for this block and the next `will_checkpoint_tensor` code?

**Comment by deepakn94** [comment]
One small comment, looks ok to me otherwise.

File: megatron/core/rerun_state_machine.py (hunk 1 above)
Comment (kwyss-nvidia) [reply megatron/core/rerun_state_machine.py]: All the boolean all reduces share a helper now.

File: megatron/core/rerun_state_machine.py [hunk 2]
```diff
@@ -249,6 +249,11 @@ def get_mode(self) -> RerunMode:
 
         return self.mode
 
+    def _reduce_any(self, value: bool) -> bool:
```
Comment (deepakn94) [inline megatron/core/rerun_state_machine.py]: Looks good, add docstring?

Comment (kwyss-nvidia) [reply megatron/core/rerun_state_machine.py]: Done.

File: megatron/core/rerun_state_machine.py [hunk 3]
```diff
@@ -347,23 +360,17 @@ def train_step(data_iterator, ...):
             return True
         # Are we done re-running from a checkpoint?
         elif self.state == RerunState.RERUNNING_FROM_CHECKPOINT:
-            will_restart_again_tensor: torch.Tensor = torch.tensor(
-                [self.restart_again_requested], dtype=torch.int32, device="cuda"
-            )
-            torch.distributed.all_reduce(will_restart_again_tensor)
-            if will_restart_again_tensor.item() > 0:
+            will_restart_again = self._reduce_any(self.restart_again_requested)
```
Comment (cyme) [inline megatron/core/rerun_state_machine.py]: Another possible improvement: perform both reductions for `will_restart_again` and `will_continue` in a single collective. _reduce_any() could accept a variable list of parameters to reduce.

Comment (deepakn94) [reply megatron/core/rerun_state_machine.py]: Yup, that's a good idea!

Comment (kwyss-nvidia) [reply megatron/core/rerun_state_machine.py]: Done. On this note, it kind of bugs me that the RERUNNING_IN_PLACE branch reduces self.continue_requested when all ranks have a consistent value, but relying on that is too brittle for my tastes, so I left a note.
